# Lab 2 - biblioteka pandas: przetwarzanie danych

Pandas jest biblioteką języka Python umożliwiającą przetwarzanie i wizualizację danych. Dzięki wygodnemu i przejrzystemu interfejsowi, jest jednym z najczęściej wykorzystywanych rozwiązań w analizie danych. 

## Instalacja

W celu instalacji biblioteki pandas należy użyć polecenia **pip install pandas**.

## Pierwsze kroki z biblioteką Pandas

Biblioteka pandas jest narzędziem zewnętrznym, więc w pierwszej kolejności należy ją zaimportować:

In [ ]:
import pandas as pd  

### Wczytywanie danych

Dane w celu ich przeanalizowania powinny zostać wczytane z dowolnego zewnętrznego źródła. Jednym z najpopularniejszych formatów transferu ustraukturyzowanych danych są pliki csv. Można je wczytać za pomocą funkcji read_csv z pakietu pandas.

In [ ]:
movie: pd.DataFrame = pd.read_csv('movie.csv')  

Dane z pliku csv zostały wczytane do obiektu klasy DataFrame, czyli ramki danych w formie tabeli. Wczytane dane można wyświetlić w łatwy sposób, przy użyciu notatnika Jupyter:

In [ ]:
movie  

color      director_name  ...  aspect_ratio  movie_facebook_likes
0     Color      James Cameron  ...          1.78                 33000
1     Color     Gore Verbinski  ...          2.35                     0
2     Color         Sam Mendes  ...          2.35                 85000
3     Color  Christopher Nolan  ...          2.35                164000
4       NaN        Doug Walker  ...           NaN                     0
...     ...                ...  ...           ...                   ...
4911  Color        Scott Smith  ...           NaN                    84
4912  Color                NaN  ...         16.00                 32000
4913  Color   Benjamin Roberds  ...           NaN                    16
4914  Color        Daniel Hsia  ...          2.35                   660
4915  Color           Jon Gunn  ...          1.85                   456

[4916 rows x 28 columns]

Ramka danych składa się z nazwanych kolumn oraz z indeksów wierszy w zakresie od 0 do 4915.

W przypadku gdy ramka danych zawiera wiele wierszy, można wykorzystać metodę head wywołaną na obiekcie klasy DataFrame, która wyświetli tylko n pierwszych wierszy:

In [ ]:
movie.head(n=3)  

color   director_name  ...  aspect_ratio  movie_facebook_likes
0  Color   James Cameron  ...          1.78                 33000
1  Color  Gore Verbinski  ...          2.35                     0
2  Color      Sam Mendes  ...          2.35                 85000

[3 rows x 28 columns]

Analogicznie metoda tail wyświetli n ostatnich wierszy:

In [ ]:
movie.tail(n=4) 

color     director_name  ...  aspect_ratio  movie_facebook_likes
4912  Color               NaN  ...         16.00                 32000
4913  Color  Benjamin Roberds  ...           NaN                    16
4914  Color       Daniel Hsia  ...          2.35                   660
4915  Color          Jon Gunn  ...          1.85                   456

[4 rows x 28 columns]

Plik movie.csv został wczytany za pomocą funkcji read_csv przy wskazaniu jedynie ścieżki. Funkcja read_csv udostępnia więcej opcji, które umożliwiają personalizację wczytywania danych. Przykładowo, plik csv jest klasycznym plikiem tekstowym, w którym wartości są odseparowane dowolnym znakiem. Pliki csv mogą również posiadać nagłówki kolumn, które nie różnią się niczym od pozostałych wierszy. Poniżej znajduje się przykład wczytania zbioru z pliku movies.csv przy jawnym wskazaniu parametrów charakteryzujących zawartość, takich jak separator w postaci przecinka, czy nagłówek w wierszu 0:

In [ ]:
movie: pd.DataFrame = pd.read_csv('/content/movie.csv', sep=',', header=0)  

Parametr sep wskazuje separator wartości w pliku csv, którego domyślną wartością jest przecinek. Argument header jest przeznaczony do wskazywania pozycji wiersza nagłówkowego w pliku csv. Jeżeli nagłówek nie istnieje, należy w parametrze header przekazać wartość None, a nazwy kolumn zostaną dopasowane automatycznie. 

### Reprezentacja kolumn oraz wierszy w ramce

W celu uzyskania listy nazw kolumn należy użyć atrybutu columns na obiekcie klasy DataFrame:

In [ ]:
movie.columns  

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

W podobny sposób można uzyskać indeksy (lub nazwy) wierszy:

In [ ]:
movie.index 

RangeIndex(start=0, stop=4916, step=1)

W przypadku zbioru filmów, wiersze są opatrzone indeksami numerycznymi w zakresie od 0 do 4915 włącznie. Istnieje jednak możliwość nadania dowolnych (w tym tekstowych) indeksów wierszom w ramce.

### Typy wartości w kolumnach ramki danych

W celu sprawdzenia typów danych, które znajdują się w poszczególnych kolumnach należy użyć atrybutu dtypes wywołanego na obiekcie DataFrame:

In [ ]:
movie.dtypes  

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

Metoda value_counts jest przeznaczona do zliczania częstości wystąpień poszczególnych wartości. Można ją z powodzeniem zastosować w podsumowaniu częstości występowania typów danych w konkretnych kolumnach:

In [ ]:
movie.dtypes.value_counts()  

float64    13
object     12
int64       3
dtype: int64

### Parametry statystyczne kolumn w ramce

Za pomocą metody describe wywoływanej na obiekcie klasy DataFrame można uzyskać informacje o podstawowych parametrach statystycznych kolumn zawierających wartości numeryczne:

In [ ]:
movie.describe()  

num_critic_for_reviews     duration  ...  aspect_ratio  movie_facebook_likes
count             4867.000000  4901.000000  ...   4590.000000           4916.000000
mean               137.988905   107.090798  ...      2.222349           7348.294142
std                120.239379    25.286015  ...      1.402940          19206.016458
min                  1.000000     7.000000  ...      1.180000              0.000000
25%                 49.000000    93.000000  ...      1.850000              0.000000
50%                108.000000   103.000000  ...      2.350000            159.000000
75%                191.000000   118.000000  ...      2.350000           2000.000000
max                813.000000   511.000000  ...     16.000000         349000.000000

[8 rows x 16 columns]

## Wybieranie kolumn

Istnieje kilka sposobów na wybór kolumn z ramki danych:

In [ ]:
col: pd.Series = movie['genres']  

col

0       Action|Adventure|Fantasy|Sci-Fi
1              Action|Adventure|Fantasy
2             Action|Adventure|Thriller
3                       Action|Thriller
4                           Documentary
                     ...               
4911                       Comedy|Drama
4912       Crime|Drama|Mystery|Thriller
4913              Drama|Horror|Thriller
4914               Comedy|Drama|Romance
4915                        Documentary
Name: genres, Length: 4916, dtype: object

In [ ]:
movie.genres 

0       Action|Adventure|Fantasy|Sci-Fi
1              Action|Adventure|Fantasy
2             Action|Adventure|Thriller
3                       Action|Thriller
4                           Documentary
                     ...               
4911                       Comedy|Drama
4912       Crime|Drama|Mystery|Thriller
4913              Drama|Horror|Thriller
4914               Comedy|Drama|Romance
4915                        Documentary
Name: genres, Length: 4916, dtype: object

Kolumny w ramkach danych opatrzone są typem Series, co oznacza że składają się z wartości opatrzonych indeksami wierszy z ramki. Oba sposoby wybierania kolumn zapewniają taki sam efekt, jednak użycie operatora kropki wyklucza wybór kolumny zawierającej w nazwie np. spację.

W podobny sposób można wybrać wiele kolumn z ramki danych przekazując listę ich nazw:

In [ ]:
movie[['actor_1_name', 'actor_2_name', 'actor_3_name']].head()  

actor_1_name      actor_2_name          actor_3_name
0      CCH Pounder  Joel David Moore             Wes Studi
1      Johnny Depp     Orlando Bloom        Jack Davenport
2  Christoph Waltz      Rory Kinnear      Stephanie Sigman
3        Tom Hardy    Christian Bale  Joseph Gordon-Levitt
4      Doug Walker        Rob Walker                   NaN

### Wybór kolumn według typu danych

Kolumny z ramki można również wybrać na podstawie charakteru danych, które się w nich znajdują. Służy do tego metoda select_dtypes wywoływana na obiekcie klasy Dataframe.

In [ ]:
movie.select_dtypes(include=['float']).head()  

num_critic_for_reviews  duration  ...  imdb_score  aspect_ratio
0                   723.0     178.0  ...         7.9          1.78
1                   302.0     169.0  ...         7.1          2.35
2                   602.0     148.0  ...         6.8          2.35
3                   813.0     164.0  ...         8.5          2.35
4                     NaN       NaN  ...         7.1           NaN

[5 rows x 13 columns]

### Wybór kolumn według frazy w nazwie

Metoda filter wywoływana na obiekcie klasy DataFrame zwróci ramkę danych z kolumnami, które zawierają wskazanę frazę za pomocą parametru like:

In [ ]:
movie.filter(like='facebook').head()  

director_facebook_likes  ...  movie_facebook_likes
0                      0.0  ...                 33000
1                    563.0  ...                     0
2                      0.0  ...                 85000
3                  22000.0  ...                164000
4                    131.0  ...                     0

[5 rows x 6 columns]

## Praca z wartościami wybrakowanymi

Niektóre wartości w kolumnach ramki mogą być wybrakowane. Metoda isnull wywołana na obiekcie klasy Series zwróci serię wartości logicznych odpowiadających temu czy dana wartośc w kolumnie jest wybrakowana czy nie:

In [ ]:
movie['genres'].isnull()  

0       False
1       False
2       False
3       False
4       False
        ...  
4911    False
4912    False
4913    False
4914    False
4915    False
Name: genres, Length: 4916, dtype: bool

Za pomocą metody any wywołanej na powstałej w ten sposób ramce można sprawdzić czy występuje tam przynajmniej jedna wartość prawdziwa:

In [ ]:
movie['genres'].isnull().any()  

False

Metoda all umożliwia sprawdzenie czy w ramce występują tylko i wyłącznie wartości prawdziwe.

Jeżeli w kolumnie występują wartości wybrakowane, można je uzupełnić wskazaną wartością, np. zerem:

In [ ]:
movie['actor_1_facebook_likes'].fillna(0)  

0        1000.0
1       40000.0
2       11000.0
3       27000.0
4         131.0
         ...   
4911      637.0
4912      841.0
4913        0.0
4914      946.0
4915       86.0
Name: actor_1_facebook_likes, Length: 4916, dtype: float64

Aby umieścić wartości wybrakowane w oryginalnej ramce danych należy użyć metody fillna na obiekcie klasy DataFrame przekazując jako parametr słownik mapujący nazwy kolumn na wartości, którymi mają zostać zastąpione wartości wybrakowane:

In [ ]:
movie.fillna({  
    'actor_1_facebook_likes': 200,
    'actor_3_facebook_likes': 100,
})  

color      director_name  ...  aspect_ratio  movie_facebook_likes
0     Color      James Cameron  ...          1.78                 33000
1     Color     Gore Verbinski  ...          2.35                     0
2     Color         Sam Mendes  ...          2.35                 85000
3     Color  Christopher Nolan  ...          2.35                164000
4       NaN        Doug Walker  ...           NaN                     0
...     ...                ...  ...           ...                   ...
4911  Color        Scott Smith  ...           NaN                    84
4912  Color                NaN  ...         16.00                 32000
4913  Color   Benjamin Roberds  ...           NaN                    16
4914  Color        Daniel Hsia  ...          2.35                   660
4915  Color           Jon Gunn  ...          1.85                   456

[4916 rows x 28 columns]

## Arytmetyka serii danych

Na obiektach klasy Series można stosować podstawowe operatory arytmetyczne. Przykładowo, do pięciu pierwszych wierszy kolumny imdb_score zostanie dodana wartość 1:

In [ ]:
movie['imdb_score'].head()  

0    7.9
1    7.1
2    6.8
3    8.5
4    7.1
Name: imdb_score, dtype: float64

In [ ]:
movie['imdb_score'].head() + 3 

0    10.9
1    10.1
2     9.8
3    11.5
4    10.1
Name: imdb_score, dtype: float64

W analogiczny sposób można dokonywać innych operacji arytmetycznych, takich jak odejmowanie, mnożenie, dzielenie.

## Operacje logiczne na serii danych

W wyniku operacji na ramkach danych, zostanie zwrócona nowa ramka zawierająca wartości logiczne będące wynikiem dokonanej operacji. Przykładowo, sprawdzenie czy 5 pierwszych wartości w kolumnie imdb_score są większe od 7.5:


In [ ]:
movie['imdb_score'].head() > 7.5  

0     True
1    False
2    False
3     True
4    False
Name: imdb_score, dtype: bool

W analogiczny sposób można dokonywać operacji logicznych porównania, większości, mniejszości itd.

## Ustawianie indeksów w ramce danych

Metoda set_index umożliwia ustawienie wskazanej kolumny jako indeks ramki:

In [ ]:
movie_new_idx: pd.DataFrame = movie.set_index('movie_title')  

movie_new_idx

color  ... movie_facebook_likes
movie_title                                        ...                     
Avatar                                      Color  ...                33000
Pirates of the Caribbean: At World's End    Color  ...                    0
Spectre                                     Color  ...                85000
The Dark Knight Rises                       Color  ...               164000
Star Wars: Episode VII - The Force Awakens    NaN  ...                    0
...                                           ...  ...                  ...
Signed Sealed Delivered                     Color  ...                   84
The Following                               Color  ...                32000
A Plague So Pleasant                        Color  ...                   16
Shanghai Calling                            Color  ...                  660
My Date with Drew                           Color  ...                  456

[4916 rows x 27 columns]

Taki sam efekt można uzyskać wskazując nazwę kolumny przy wczytywaniu pliku csv (funkcja read_csv) w parametrze index_col.

## Zmiana nazw kolumn oraz indeksów wierszy

W tym celu należy użyć metody rename wywołanej na obiekcie klasy DataFrame przekazując słowniki mapujące stare nazwy na nowe:

In [ ]:
from typing import Dict

idx_rename: Dict[str, str] = {  
    'Avatar': 'Avatar II',
    'Spectre': 'Spectre II',
}

col_rename: Dict[str, str] = {  
    'director_facebook_likes': 'Director FB likes',
    'num_critic_for_reviews': 'Num critic for reviews',
}

movie_new_idx = movie_new_idx.rename(index=idx_rename, columns=col_rename)  

movie_new_idx.head(n=7)  

color  ... movie_facebook_likes
movie_title                                        ...                     
Avatar II                                   Color  ...                33000
Pirates of the Caribbean: At World's End    Color  ...                    0
Spectre II                                  Color  ...                85000
The Dark Knight Rises                       Color  ...               164000
Star Wars: Episode VII - The Force Awakens    NaN  ...                    0
John Carter                                 Color  ...                24000
Spider-Man 3                                Color  ...                    0

[7 rows x 27 columns]

## Tworzenie nowych kolumn

Kolumnę ze stałymi wartościami (np. 100) można dodać następująco: (kolumna znajdzie się na końcu ramki)

In [ ]:
movie_new_idx['visits_num'] = 100  

movie_new_idx.head()

color  ... visits_num
movie_title                                        ...           
Avatar II                                   Color  ...        100
Pirates of the Caribbean: At World's End    Color  ...        100
Spectre II                                  Color  ...        100
The Dark Knight Rises                       Color  ...        100
Star Wars: Episode VII - The Force Awakens    NaN  ...        100

[5 rows x 28 columns]

Biblioteka pandas umożliwia również tworzenie nowych kolumn w ramce na podstawie istniejących. Przykładowo, będzie to kolumna sumująca wszystkie lajki uzyskane przez aktorów i reżysera:

In [ ]:
movie_new_idx['total_facebook_likes'] = (movie_new_idx['actor_1_facebook_likes'] + movie_new_idx['actor_2_facebook_likes'] + movie_new_idx['actor_3_facebook_likes'] + movie_new_idx['Director FB likes'])

movie_new_idx.head() 

color  ... total_facebook_likes
movie_title                                        ...                     
Avatar II                                   Color  ...               2791.0
Pirates of the Caribbean: At World's End    Color  ...              46563.0
Spectre II                                  Color  ...              11554.0
The Dark Knight Rises                       Color  ...              95000.0
Star Wars: Episode VII - The Force Awakens    NaN  ...                  NaN

[5 rows x 29 columns]

## Usuwanie kolumn

Do usuwania kolumn z ramki danych można wykorzystać instrukcję del:

In [ ]:
del movie_new_idx['color']  

movie_new_idx.head()

director_name  ...  total_facebook_likes
movie_title                                                    ...                      
Avatar II                                       James Cameron  ...                2791.0
Pirates of the Caribbean: At World's End       Gore Verbinski  ...               46563.0
Spectre II                                         Sam Mendes  ...               11554.0
The Dark Knight Rises                       Christopher Nolan  ...               95000.0
Star Wars: Episode VII - The Force Awakens        Doug Walker  ...                   NaN

[5 rows x 28 columns]

## Wybór wierszy

Za pomocą atrybutu tablicowego loc ramki danych można wskazać wiersz lub wiersze do wyboru wskazując indeks:

In [ ]:
movie_new_idx.loc['Avatar II']

director_name                                                    James Cameron
Num critic for reviews                                                     723
duration                                                                   178
Director FB likes                                                            0
actor_3_facebook_likes                                                     855
actor_2_name                                                  Joel David Moore
actor_1_facebook_likes                                                    1000
gross                                                              7.60506e+08
genres                                         Action|Adventure|Fantasy|Sci-Fi
actor_1_name                                                       CCH Pounder
num_voted_users                                                         886204
cast_total_facebook_likes                                                 4834
actor_3_name                                        

In [ ]:
movie_new_idx.loc[['Avatar II', 'Spectre II']]  

director_name  ...  total_facebook_likes
movie_title                 ...                      
Avatar II    James Cameron  ...                2791.0
Spectre II      Sam Mendes  ...               11554.0

[2 rows x 28 columns]

Jeżeli ramka danych posiada indeks numeryczny, w atrybucie loc wystarczy wskazać numeryczne pozycje wierszy:

In [ ]:
movie.loc[[0, 1, 2]]  

color   director_name  ...  aspect_ratio  movie_facebook_likes
0  Color   James Cameron  ...          1.78                 33000
1  Color  Gore Verbinski  ...          2.35                     0
2  Color      Sam Mendes  ...          2.35                 85000

[3 rows x 28 columns]

Atrybut tablicowy iloc (podobnie jak w bibliotece NumPy) służy do wybierania wierszy i kolumn wskazując ich pozycje numerycznie:

In [ ]:
movie_new_idx.iloc[0:10, :]  

director_name  ...  total_facebook_likes
movie_title                                                    ...                      
Avatar II                                       James Cameron  ...                2791.0
Pirates of the Caribbean: At World's End       Gore Verbinski  ...               46563.0
Spectre II                                         Sam Mendes  ...               11554.0
The Dark Knight Rises                       Christopher Nolan  ...               95000.0
Star Wars: Episode VII - The Force Awakens        Doug Walker  ...                   NaN
John Carter                                    Andrew Stanton  ...                2277.0
Spider-Man 3                                        Sam Raimi  ...               39000.0
Tangled                                          Nathan Greno  ...                1651.0
Avengers: Age of Ultron                           Joss Whedon  ...               66000.0
Harry Potter and the Half-Blood Prince            David Yates  ...               46282.0

[10 rows x 28 columns]

In [ ]:
movie_new_idx.iloc[:, 3:7]  

Director FB likes  ...  actor_1_facebook_likes
movie_title                                                    ...                        
Avatar II                                                 0.0  ...                  1000.0
Pirates of the Caribbean: At World's End                563.0  ...                 40000.0
Spectre II                                                0.0  ...                 11000.0
The Dark Knight Rises                                 22000.0  ...                 27000.0
Star Wars: Episode VII - The Force Awakens              131.0  ...                   131.0
...                                                       ...  ...                     ...
Signed Sealed Delivered                                   2.0  ...                   637.0
The Following                                             NaN  ...                   841.0
A Plague So Pleasant                                      0.0  ...                     0.0
Shanghai Calling                                          0.0  ...                   946.0
My Date with Drew                                        16.0  ...                    86.0

[4916 rows x 4 columns]

## Zadania

1. Wczytać zbiór danych z pliku movie_new.csv do ramki movie_new. Przy użyciu parametru names funkcji read_csv nadać następujące nazwy kolumnom: color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,actor_1_name,movie_title,num_voted_users,cast_total_facebook_likes,actor_3_name,facenumber_in_poster,plot_keywords,movie_imdb_link,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes.

2. Utworzyć ramkę movie_num, w której będą zawarte jedynie kolumny z wartościami typu numerycznego.

3. Wypełnić brakujące wartości w ramce movie_num średnią arytmetyczną wyznaczoną dla każdej kolumny z osobna. 

4. Znaleźć kolumnę w ramce movie_num, której wartości mają największe odchylenie standardowe.

5. Zastąpić kolumny numeryczne w ramce movie_new nowymi kolumnami z ramki movie_num.

6. Zastąpić wszystkie wartości wybrakowane w ramce movie_new za pomocą pierwszej wartości występującej w kolumnie.

7. Nadać nowe indeksy wierszom w ramce movie_new według następującego schematu: [imie i nazwisko rezysera] - [tytul filmu].

8. Zwiększyć wartość kolumny imdb_score każdego filmu w ramce movie_new o 1.

9. Zmienić nazwy wszystkich kolumn w ramce movie_new w taki sposób aby zlikwidować wszystkie znaki _ oraz pierwsze litery wszystkich słów w nazwach były wielkie.

10. Utworzyć ramkę movie_random, w której zostaną wybrane wiersze o losowej ilości w zakresie pozycji [start, end], gdzie start i end będą liczbami pseudolosowymi.

11. Zwiększyć wartości każdej kolumny numerycznej w ramce movie_random o wartość pseudolosową (po jednej wartości na kolumnę).